src/NHibernate/Impl/SessionImpl.cs

```diff
@@ -2604,5 +2608,10 @@ public override IEntityPersister GetEntityPersister(string entityName, object ob
 				}
 			}
 		}
+
+		void RemoveChildSession(ISession session)
+		{
+			Interlocked.CompareExchange(ref _childSession, null, session);
```

Comment:
Just seeing know these [commit comments](https://github.com/nhibernate/nhibernate-core/commit/390011bca21bf90bb6912d62d97c229de3ec26cf#commitcomment-21874582) on the same subject.

In do not really get how another child-session could be there with current implementation. But well, if it is not about thread-safety but only for not removing another one, I would rather implement it without any threading helper, with a if or iif.